canonical/cos-lite-bundle
# Add tls overlay

This PR adds a TLS overlay.

Fixes #78.

Depends on:
- https://github.com/canonical/traefik-k8s-operator/pull/204
- https://github.com/canonical/catalogue-k8s-operator/issues/31
- https://github.com/canonical/alertmanager-k8s-operator/issues/189
- https://github.com/canonical/prometheus-k8s-operator/issues/511
- https://github.com/canonical/grafana-k8s-operator/pull/257
- https://github.com/canonical/prometheus-k8s-operator/pull/524
- https://github.com/canonical/alertmanager-k8s-operator/pull/191
- https://github.com/canonical/traefik-k8s-operator/pull/245
- https://github.com/canonical/observability-libs/pull/60
- https://github.com/canonical/tls-certificates-interface/pull/74
- https://github.com/canonical/traefik-k8s-operator/pull/249
- https://github.com/canonical/prometheus-k8s-operator/issues/533

## Testing
First, deploy the testing overlay:
```
$ tox -e render-edge
$ juju deploy ./bundle.yaml --trust \
  --overlay overlays/tls-overlay.yaml \
  --overlay overlays/testing-overlay.yaml
```

We do not have the `receive-ca-cert` relation yet in cos charms, so the external ca would need to be copied in manually:
```
juju run external-ca/0 get-ca-certificate --format json | jq -r '."external-ca/0".results."ca-certificate"' > external-ca.crt
openssl x509 -noout -text -in external-ca.crt

echo | openssl s_client -showcerts -servername 10.43.8.206 -connect 10.43.8.206:443 | openssl x509 -text -noout

juju scp --container prometheus external-ca.crt prometheus/0:/usr/local/share/ca-certificates
juju ssh --container prometheus prometheus/0 update-ca-certificates --fresh

juju scp --container grafana external-ca.crt grafana/0:/usr/local/share/ca-certificates
juju ssh --container grafana grafana/0 update-ca-certificates --fresh
```

Then:
1. Attempt to manually `curl` each cos component: `juju show-unit catalogue/0 | grep url`.
2. Make sure all scrape targets are healthy: `curl -k https://10.43.8.206/bndl-prometheus-0/api/v1/targets | jq | grep -E "scrapeUrl|health"`.
3. Make sure the watchdog alert from avalanche is firing:
    - In prometheus: `curl <prometheus>/api/v1/alerts`.
    - In alertmanager: `curl <alertmanager>/api/v2/alerts`.
4. Make sure grafana has all the dashboards.
